bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/service/RemoteControllerService.java

```diff
@@ -215,6 +215,9 @@ public void start() {
         } else if (SamsungTvConfiguration.PROTOCOL_WEBSOCKET.equals(protocol)
                 || SamsungTvConfiguration.PROTOCOL_SECUREWEBSOCKET.equals(protocol)) {
             try {
+                for (EventListener listener : listeners) {
+                    listener.putOffline();
```

Comment:
Why should it be set OFFLINE even before trying? Isn't the websocket only tried to establish in the next line?

Reply:
From what I can tell, on legacy TVs, reportError executed is ultimately what sets the TV to OFFLINE at start because an error is thrown.  On WebSocket TVs, no error is thrown so the device is never put into an OFFLINE state and initialized as such.

I believe that the legacy connection throws the error because socket.connect() fails here.

https://github.com/openhab/openhab-addons/blob/25498701bf0ae67c84ceebfe62ecc63cd841c847/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/service/RemoteControllerService.java#L231

 https://github.com/openhab/openhab-addons/blob/25498701bf0ae67c84ceebfe62ecc63cd841c847/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerLegacy.java#L95
  
However on WebSocket, client.start() doesn't fail and throw the error.

https://github.com/openhab/openhab-addons/blob/25498701bf0ae67c84ceebfe62ecc63cd841c847/bundles/org.openhab.binding.samsungtv/src/main/java/org/openhab/binding/samsungtv/internal/protocol/RemoteControllerWebSocket.java#L147

Without putting the device offline at this point, it never goes offline until the TV is turned on for the first time.  This causes issues with rules because things like power stay in a bad state until that point.  When comparing to legacy clients, unless the device is actually online, it will always become offline after the connection attempt is made.  Given that offline is always the end state (minus online), I didn't see a negative to putting it in an offline state first and then let the rest of the code decide if the TV is actually online.

Reply:
> I didn't see a negative to putting it in an offline state first and then let the rest of the code decide if the TV is actually online.

No, this is definitely something that must be avoided. You must never set a wrong status, so as long as it cannot be determined, it has to stay UNKNOWN.
If the websocket connection does not end in an error, can't the status be set to ONLINE after the connect statement?
Are you actually sure that it cannot be established? Even if powered off, my TV should still be ONLINE, since it supports art mode, which is also correctly reported to openHAB while being off.

Reply:
The older websocket TVs don't do that. They completely shut the network stack down after being off for 5-10 minutes. No pings, nothing. If you send a WoL packet, the TV powers up. It's a very annoying thing that has only been fixed recently.  All 3 of my TVs are OFFLINE and completely unresponsive unless actually ON.

Reply:
Above you said that legacy TVs throw the websocket error, so they should go OFFLINE as expected, right?
And if the newer TVs establish the websocket connection, the thing should simply go to ONLINE as I suggested above.

Reply:
Sorry, let me be more clear. There are 3 types of Samsung TVs that this binding has worked with. Legacy is an older protocol that is not WS based. That protocol goes to OFFLINE when the error above is thrown. There is then a generation of WebSocket (which is what I have that is going to UNKNOWN) that this is intended to fix. There is then the newest generation (which I belive you have as you mentioned you just got yours) which seems to behave better. I believe this is considered SecureWebSocket and doesn't have this issue.

Reply:
Ok, so what exactly happens with the websocket connection of those "mid-generation" TVs then? You say it is not established, but no error is thrown either? How can that happen? I would expect the websocket client to clearly notice whether there's a connection error or not and report this correctly. Hence if there's no error, I assume the connection is established for them as well and hence the thing should go ONLINE.

Reply:
Then perhaps we should check the WS code. When these TVs are not on, they are completely unresponsive. I suppose you could simulate by creating a thing to an IP that has nothing on it. When the connection fails to establish, no error is caught and the TV stays UNKNOWN.

Reply:
> When these TVs are not on, they are completely unresponsive.

I'd then recommend to send some other event to them through the websocket to see whether they respond or not. Based on that, the status can be set to either ONLINE or OFFLINE.

Reply:
I'm not opposed to that.  Should we also figure out why the WS code isn't throwing an error when the connection fails?

Reply:
Well, first figure out whether the connection REALLY fails. Maybe the connection works, but the TV simply does not react upon any messages through that connection.

Reply:
The connection is 1000% not connecting.  The mac address for the TV isn't even in my switch at that point, there is absolutely no way that anything is connecting to it.  I've tried both wired and wifi, same results.  On wifi, I can actually see the client disconnect from the network.  This is what OH sees from my TV being turned on and off using the GlobalCache IR that I use.  Notice that the "networksamsung" items (which are attached to the network binding) show the TV begin responding to pings and then cutting pings off after the TV is turned off.

```
2021-01-24 07:22:52.261 [INFO ] [openhab.event.ItemCommandEvent      ] - Item 'GC000C1E04A4F7_M1C1_Button' received command SAMSUNGTV_POWER_TOGGLE
2021-01-24 07:22:52.261 [INFO ] [openhab.event.ItemCommandEvent      ] - Item 'GC000C1E04A4F7_M1C1' received command SAMSUNGTV_POWER_TOGGLE
2021-01-24 07:22:52.265 [INFO ] [penhab.event.ItemStatePredictedEvent] - Item 'GC000C1E04A4F7_M1C1' predicted to become SAMSUNGTV_POWER_TOGGLE
2021-01-24 07:22:52.268 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_PowerState' changed from OFF to ON
2021-01-24 07:22:52.268 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'GC000C1E04A4F7_M1C1' changed from EMPTY to SAMSUNGTV_POWER_TOGGLE
2021-01-24 07:22:52.518 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_PowerString' changed from OFF to ON/OFFLINE
2021-01-24 07:22:53.130 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4online' changed from OFF to ON
2021-01-24 07:22:53.130 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from UNDEF to 1.0
2021-01-24 07:22:54.117 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:18f315ed_28c8_4066_a4a4_d3cadd59a670' changed from OFFLINE to ONLINE
2021-01-24 07:22:54.117 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:18f315ed_28c8_4066_a4a4_d3cadd59a670' changed from ONLINE to OFFLINE
2021-01-24 07:22:54.122 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_Power' changed from OFF to ON
2021-01-24 07:22:54.122 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_Power' changed from ON to OFF
2021-01-24 07:22:54.175 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:18f315ed_28c8_4066_a4a4_d3cadd59a670' changed from OFFLINE to ONLINE
2021-01-24 07:22:54.175 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_Power' changed from OFF to ON
2021-01-24 07:22:54.180 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:18f315ed_28c8_4066_a4a4_d3cadd59a670' changed from ONLINE to OFFLINE
2021-01-24 07:22:54.180 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_Power' changed from ON to OFF
2021-01-24 07:22:54.188 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:18f315ed_28c8_4066_a4a4_d3cadd59a670' changed from OFFLINE to ONLINE
2021-01-24 07:22:54.188 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_Power' changed from OFF to ON
2021-01-24 07:22:55.250 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:18f315ed_28c8_4066_a4a4_d3cadd59a670' changed from ONLINE to OFFLINE
2021-01-24 07:22:55.251 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_Power' changed from ON to OFF
2021-01-24 07:22:55.265 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:18f315ed_28c8_4066_a4a4_d3cadd59a670' changed from OFFLINE to ONLINE
2021-01-24 07:22:55.266 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_Power' changed from OFF to ON
2021-01-24 07:22:55.432 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 1.0 to 2.0
2021-01-24 07:22:59.535 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 2.0 to 1.0
2021-01-24 07:23:03.638 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 1.0 to 2.0
2021-01-24 07:23:09.794 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 2.0 to 1.0
2021-01-24 07:23:11.846 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 1.0 to 2.0
2021-01-24 07:23:18.001 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 2.0 to 1.0
2021-01-24 07:23:20.053 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 1.0 to 2.0
2021-01-24 07:23:24.376 [INFO ] [openhab.event.ItemCommandEvent      ] - Item 'GC000C1E04A4F7_M1C1_Button' received command SAMSUNGTV_POWER_TOGGLE
2021-01-24 07:23:24.376 [INFO ] [openhab.event.ItemCommandEvent      ] - Item 'GC000C1E04A4F7_M1C1' received command SAMSUNGTV_POWER_TOGGLE
2021-01-24 07:23:24.377 [INFO ] [penhab.event.ItemStatePredictedEvent] - Item 'GC000C1E04A4F7_M1C1' predicted to become SAMSUNGTV_POWER_TOGGLE
2021-01-24 07:23:24.379 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_PowerState' changed from ON to OFF
2021-01-24 07:23:24.380 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'GC000C1E04A4F7_M1C1' changed from SAMSUNGTV_POWER_TOGGLE to EMPTY
2021-01-24 07:23:26.131 [INFO ] [ab.event.ThingStatusInfoChangedEvent] - Thing 'samsungtv:tv:18f315ed_28c8_4066_a4a4_d3cadd59a670' changed from ONLINE to OFFLINE
2021-01-24 07:23:26.132 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'SamsungtvTv18f315ed_28c8_4066_a4a4_d3cadd59a670_Power' changed from ON to OFF
2021-01-24 07:23:26.208 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 2.0 to 1.0
2021-01-24 07:23:28.262 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 1.0 to 4.0
2021-01-24 07:23:30.311 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 4.0 to 2.0
2021-01-24 07:23:34.415 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 2.0 to 1.0
2021-01-24 07:23:36.466 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 1.0 to 2.0
2021-01-24 07:23:40.570 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 2.0 to 1.0
2021-01-24 07:23:45.272 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4online' changed from ON to OFF
2021-01-24 07:23:45.272 [INFO ] [openhab.event.ItemStateChangedEvent ] - Item 'networksamsunga4a4time' changed from 1.0 to UNDEF


```

Reply:
@kaikreuzer I went through the WS logic on the binding and I have a proposal for a way to do this better.  openConnection() calls connectWebSockets() after the try/catch.  We could put an if statement at the end of connectWebSockets() to check all 3 sockets using isConnected() and if they aren't connected at that point we could call reportError().  Or, option 2, we could have connectWebSockets() return true/false based on success and have openConnection() fire reportError() if it returns false.  Or, option 3, move connectWebSockets() into the try/catch in openConnection() and have it throw an error that the catch in openConnection() will report back.  All basically do the same thing, just preference on how to execute it.

Reply:
Or none of those will work...  Apparently there is a timing issue here with the connection.  I added:

`        logger.info("Check WS {} {} {}",webSocketRemote.isConnected(),webSocketArt.isConnected(),webSocketV2.isConnected());`

at the end of connectWebSockets() to see what it would kick back.  I assumed I would get some or all of them replying true.  Notice that it throws the message to the log before the websockets connect below.

```
2021-01-24 11:07:54.862 [INFO ] [rnal.service.RemoteControllerService] - Using WebSocket interface
2021-01-24 11:07:54.864 [DEBUG] [l.protocol.RemoteControllerWebSocket] - RemoteControllerWebSocket start Client
2021-01-24 11:07:54.866 [DEBUG] [ngtv.internal.protocol.WebSocketBase] - WebSocketRemote connecting to: ws://REDACTED.153:8001/api/v2/channels/samsung.remote.control?name=REDACTED==
2021-01-24 11:07:54.867 [DEBUG] [ngtv.internal.protocol.WebSocketBase] - WebSocketArt connecting to: ws://REDACTED.153:8001/api/v2/channels/com.samsung.art-app?name=REDACTED==
2021-01-24 11:07:54.867 [DEBUG] [ngtv.internal.protocol.WebSocketBase] - WebSocketV2 connecting to: ws://REDACTED.153:8001/api/v2?name=REDACTED==
2021-01-24 11:07:54.867 [INFO ] [l.protocol.RemoteControllerWebSocket] - Check WS false false false
2021-01-24 11:07:54.867 [INFO ] [l.protocol.RemoteControllerWebSocket] - Check WS false false false
2021-01-24 11:07:54.878 [DEBUG] [ngtv.internal.protocol.WebSocketBase] - WebSocketArt connection established: REDACTED.153
2021-01-24 11:07:54.878 [DEBUG] [ngtv.internal.protocol.WebSocketBase] - WebSocketRemote connection established: REDACTED.153
2021-01-24 11:07:55.026 [DEBUG] [ngtv.internal.protocol.WebSocketBase] - WebSocketRemote connection error
2021-01-24 11:07:55.026 [DEBUG] [ungtv.internal.protocol.WebSocketArt] - Art client connected
2021-01-24 11:07:55.026 [DEBUG] [ungtv.internal.protocol.WebSocketArt] - Art client connected
2021-01-24 11:07:55.027 [DEBUG] [ngtv.internal.protocol.WebSocketBase] - WebSocketRemote connection closed: 1002 - Frame forbidden close status code: 1005
2021-01-24 11:07:55.027 [DEBUG] [ngtv.internal.protocol.WebSocketBase] - WebSocketV2 connection established: REDACTED.153
2021-01-24 11:07:55.027 [DEBUG] [ungtv.internal.protocol.WebSocketArt] - Art channel connected
2021-01-24 11:07:55.045 [DEBUG] [sungtv.internal.protocol.WebSocketV2] - V2 client connected
2021-01-24 11:07:55.045 [DEBUG] [sungtv.internal.protocol.WebSocketV2] - V2 client connected
2021-01-24 11:07:55.046 [DEBUG] [sungtv.internal.protocol.WebSocketV2] - V2 channel connected. Token = null

```